### ISSUE TITLE
Kyrian MM Hunter Rotation APL seems very off. (Especially Trueshot Rotation)

### ISSUE THREAD

**Issue by biehlieh** [Hekili/hekili#1149]
1.  Check to see if you are using the latest version of the addon.  If there is a newer alpha/beta/release version, see if your problem exists with the new version.

I am using v9.1.0-1.0.7b

2.  Check to see if someone else has already opened an issue report here.  Please don't submit repeat reports.

I did not find anything regarding kyrian MM hunter, only survival issues.

**Describe the Bug/Issue**

The APL/Rotation Suggestions for Kyrian Marksmanship Hunters seems very wrong, especially during Trueshot. 
The Opener according to the Icy-Veins guide (written by Azortharion, a well-respected pro-Hunter) should be:
1 Cast Double Tap around 8 seconds before combat begins.
2 Start casting Steady Shot around 2 seconds before combat starts.
3 Cast an additional Steady Shot to activate Steady Focus.
4 Cast Resonating Arrow. Activate all on-use trinkets and Racials here as well.
5 Cast Trueshot.
6 Cast Aimed Shot.
7 Cast Rapid Fire.
8 Cast Aimed Shot

As it stands, Hekili suggests:
1. starting with Aimed Shot, 
2. double Steady shot,
3.  Double Tap, 
4. Resonating Arrow, 
5. Trueshot.
6. Rapid fire
7. Aimed Shot

and will continue with a very weird usage of Aimed Shot e.g. letting Aimed Shot reach 2 Stacks and suggesting Rapid Fire while at max Focus.

See my Video/gif. In the gif, you can also see some weird behavior concerning Lock'n'Load procs.
Attached gif so you can see the rotation issues: https://im2.ezgif.com/tmp/ezgif-2-68f94f38a5bc.gif  i hope it works.

**To Reproduce**
- Go to Training dummies and practice the Rotation as it is suggested by hekili.

**Expected behavior**
I expected the rotation APL to be correct, by maximising Aimed Shot usage as well as not overcapping focus or using rapid fire while at max focus for example.

**Screenshots**
i have created a gif to better understand the issues. https://im2.ezgif.com/tmp/ezgif-2-68f94f38a5bc.gif

**WHEN SUBMITTING AN ISSUE:**

Please provide the following information.  Leaving it out may result in your ticket being ignored (because I won't have the information needed to explore your report).

1.  The information from the Issue Reporting tab ( /hekili > Issue Reporting ).  This provides your talent and gear information.  You can paste it to pastebin.com and provide a link here.

https://pastebin.com/Vvb33AGM

2.  If reporting a disagreement with the addon's recommendations, please take a snapshot when you observe the issue and provide the snapshot.  This is not a screenshot.  A snapshot can be generated by binding the Pause feature on the Toggles tab.  When you press Pause, a snapshot of how the addon decided on its CURRENT recommendations is generated and saved to the Snapshots tab.  Copy and paste this to pastebin.com and provide the link here.
 
#1 https://pastebin.com/S1X3jHu2
#2 https://pastebin.com/uJDFW4hW
#3 https://pastebin.com/kJbHNAn7

3.  If reporting an error (i.e., the addon freezes and/or disables itself), please install the BugSack and BugGrabber addons as they will enable you to paste the exact error message, with detail that will help me identify where the code error is located.  You can paste the error information here or use pastebin.com for that as well.

**PUT ERROR MESSAGE(S) HERE:**

**Comment by Hekili** [comment]
Compare to your sim, not to a guide.  Can you link your sim?

**Comment by biehlieh** [comment]
https://www.raidbots.com/simbot/report/2RHYG5ULJDs1tTPwoENTVJ

this is my sim. while yes, the opener seems to be somewhat like in hekili (double steady, then double tap into cooldowns) it still uses aimed shot after entering trueshot.

**Comment by Hekili** [comment]
> The Opener according to the Icy-Veins guide (written by Azortharion, a well-respected pro-Hunter) should be:

It's becoming a bit more common for sim behavior and what people write in guides not to match up 1:1.  My addon design is based on following what's proven in SimC.  Azor could update the priority in SimC or have someone do it if his guide is more performant.

> while yes, the opener seems to be somewhat like in hekili (double steady, then double tap into cooldowns) it still uses aimed shot after entering trueshot.

The sim priority will recommend Double Tap pre-combat if it expects there to be multiple targets -- which the addon won't do unless you are using a display mode that includes AOE.  Because the addon otherwise doesn't know there are multiple targets until you hit them.  (Pre-combat double-tap is also used for non-Kyrians without Volley talented, but that's not you.)

Here's why Aimed Shot was recommended during Trueshot for you.  First, nothing higher in the priority had its criteria met.  Second:

```
 - this entry's criteria PASSES: buff.precise_shots.down[false] | ( buff.trueshot.up[true] | full_recharge_time[0.00] < gcd.execute[1.23] + cast_time[0.00] ) & ( ! talent.chimaera_shot.enabled[false] | active_enemies[1.00] < 2 ) | buff.trick_shots.remains[0.00] > execute_time[1.23] & active_enemies[1.00] > 1
```

The relevant part is `( buff.trueshot.up[true] | full_recharge_time[0.00] < gcd.execute[1.23] + cast_time[0.00] ) & ( ! talent.chimaera_shot.enabled[false] | active_enemies[1.00] < 2 )`.

The first parentheses is true -- Trueshot is up, so we can disregard the latter half (but it's true, as well; your Aimed would be fully recharged in the time it takes you to cast Aimed Shot, because it's already fully recharged).

The second parentheses is also true, both because you don't have Chimaera Shot talented and because you have a single enemy target.

This entry is also meant to support your SST legendary.

Here is your sim action.  The Aimed Shot entries labeled `N` are the same entry that is showing in Hekili:

![image](https://user-images.githubusercontent.com/9438134/124921905-eac17e00-dfc6-11eb-8144-ad3196e95015.png)

You can see that usage of Aimed Shot during Trueshot is happening quite a bit in sims.  If that's a mistake, you might want to discuss with Putro or other SimC hunters to see if changes need to be made there.

If I missed any other issues here, please let me know.  As it stands, this appears to be working as intended/designed.

**Comment by Hekili** [comment]
Sorry, trying to pull out any unresolved issues you noted -- it's really helpful if you make a bulleted list rather than sprinkling them throughout.  You did do a great job of supplying data, so I appreciate it.

> In the gif, you can also see some weird behavior concerning Lock'n'Load procs.

Please be more specific and describe the "weird behavior."

> very weird usage of Aimed Shot e.g. letting Aimed Shot reach 2 Stacks and suggesting Rapid Fire while at max Focus.

The example here is useful, "very weird" is less useful.  The way to see why this is happening is to take a snapshot when you think the addon should be recommending Aimed Shot before reaching 2 charges but isn't.  Then I can say "it recommended a higher priority action" or "there appears to be a bug" or whatever the case may be.

> 1 Cast Double Tap around 8 seconds before combat begins.
> 2 Start casting Steady Shot around 2 seconds before combat starts.

For 1, the addon has no means of knowing you're going to start combat 8 seconds from now (btw, this is 10 seconds pre-combat in SimC).

For 2, the precombat Aimed Shot is this one:

> actions.precombat+=/aimed_shot,if=active_enemies<3&(!covenant.kyrian&!talent.volley|active_enemies<2)&!variable.etf_precast

This is subject to the same target-counting design that I mentioned earlier.  The addon can't see that there are multiple targets for a ranged spec until you hit them.  Unless you use a display mode that includes AOE, where the target count is forced.

So far, everything the addon is doing seems to match the sim behavior to the extent possible.  Keep in mind, there will be incidental differents with each pull depending on whether LnL procs and so forth.

**Comment by biehlieh** [comment]
Thank you for your very detailed answer. Sorry for being so vague in my descriptions, it is my first time contacting an addon dev :) plus i am not used to github.

> Here's why Aimed Shot was recommended during Trueshot for you. First, nothing higher in the priority had its criteria met. 
> 
Sadly my gif seems to be taken down, if you want, i can try to upload another one. Let me try to be more descriptive with "weird behaviour": 

- after activating Resonating Arrow and Trueshot, i am at full focus. this is where i would expect hekili to suggest using aimed shot, which it doesn't; it is suggesting rapid fire at full focus while also leaving aimed shot at two stacks. Rapid fire has a relatively low priority during trueshot, while spending aimed shot chargest is probably the highest. 
-  the "very weird behaviour" concerning lock'n'load: i am getting a proc at about 15 focus, hekili does not suggest using the proc right away, which is fine. The more "problematic" part is the suggestion of casting two steady shots to get steady focus up, causing me to regen focus and getting me to roughly 65 focus, but then recommending rapid fire, which brings me to 100 focus right as the cast finishes and THEN using the lock and load proc (which causes aimed shot to cost no focus, essentially leaving me at 100 focus for several seconds.

> This entry is also meant to support your SST legendary.
> 

Sorry if i am being vague again here, but it "feels" like the APL seems to be cut for the surging shots legendary.

> Here is your sim action. The Aimed Shot entries labeled `N` are the same entry that is showing in Hekili:
> 
> ![image](https://user-images.githubusercontent.com/9438134/124921905-eac17e00-dfc6-11eb-8144-ad3196e95015.png)
> 
> You can see that usage of Aimed Shot during Trueshot is happening quite a bit in sims. If that's a mistake, you might want to discuss with Putro or other SimC hunters to see if changes need to be made there.
> 

I would very much expect the usage of aimed shot to be like it is in the sim, which is why i originally posted this ticket, sorry if i wasn't clear enough.

To summarize and emphasize: 
- My sim on raidbots as well as your script  shows the behaviour i would expect of the usage of aimed shot
- The recommendations in your Addon do not seem to line up with those results (using rapid fire at full focus during trueshot, while 2 charges of aimed shot are available)
- It worked completely perfect with night fae
- The Issue seems to be only related to the rotation during trueshot.

I hope i could get my point across a bit better and i thank you for all your work!

**Comment by Hekili** [comment]
You've done well.  Can you provide specific snapshots where you're having the specific issue you describe?  For example:

> (using rapid fire at full focus during trueshot, while 2 charges of aimed shot are available)

Does one of your snapshots show this?

**Comment by biehlieh** [comment]
Hey, thanks for the quick reply.

i have added some more pastebins (i took a few snapshots in the timeframe of about 1 second, just to be safe)
- https://pastebin.com/dG7QGBsa  
- https://pastebin.com/fGbkqBQ8
- https://pastebin.com/Karw7Uyn


plus an image of the rotation recommendation before activating the cooldowns: 

- https://imgur.com/a/LNE9jD1

I hope it is easy to see that i am at 100 focus there and that the first action after activating trueshot is rapid fire.

Thanks again!

**Comment by Hekili** [comment]
It looks like there are some significant differences in the SimC MM APL that I didn't catch for the addon's translation.  I'll post a beta build for you to test and let me know if it makes any difference.

**Comment by biehlieh** [comment]
Very cool, Thanks!

**Comment by Hekili** [comment]
Okay, 1.0.8-beta1 is being built right now.  If you could test, that would be great.  The best way to do that is like you did most recently -- a bullet for each remaining issue with the snapshot showing that issue immediately underneath it.

* Issue 1
  * Snapshot 1
* Issue 2
  * Snapshot 2

Thanks for your thorough reporting.

**Comment by biehlieh** [comment]
Hi again!

I have tried it and it seems like the issues have been solved. It now starts the trueshot rotation with Aimed Shot and follows up with Rapid fire and Arcane/Aimed shot.

If you want, i can still send you a snapshot of how it works now, otherwise i think my issue has been solved.

Thanks so much!

**Comment by Hekili** [comment]
Nope, if there's no problem, no more info is needed.  Thanks for reporting, following up, and testing so quickly.  Take care!
